### ISSUE TITLE
Redis with TLS has issues with missing IP addresses in cert.

### ISSUE THREAD

**Issue by opencols** [StackExchange/StackExchange.Redis#2287]
Hi all,

Was hoping that somebody could help.

We are using Redis (Community 6.2.6) and Stack Exchange 2.6.48
Firstly, I do understand that there are several what I could call “nearly” issues available, but I have not come across a solution so I though I would put it all down.
Some background:

We create three VMs and install Redis and Redis sentinel on all the nodes.  This is done via a simple ansible script which does a compile of the tar.gz file and caters for TLS.  In the same script, we also create a self-signed .crt and .key combination and plug them into the redis.conf and sentinel.conf files.  Each VM has its own cert.  The resulting instance is (one master, two replicas and three Sentinels.  Just FYI this works perfectly from Redis Insight (2) and from a simple Python script.  We can issue CRUD commands, info etc. and replication works fine.  In addition, failover also works well.  Again, for the sake of clarity we have not tested this “self-signed” version with .net and  Stack exchange.
The next step is to replace the self-signed certs with fully signed ones.  There is a departure here in that we create just one server cert (we are not using client certs) with one name and six subject alternate names these being the IP address and FQDNof each node.
After plugging in the .crt and .key we restart Redis and Sentinel (obviously with the relevant parameters changed.  This works perfectly from Redis Insight, Python and from a .net implementation so all was good with the world but then 😊
We had created several instances in this manner when the crypto team decided that we are not allowed to use IP addresses as SANs.  We therefore decided to try and create the cert without IP addresses.  Once again, the instance continued to work perfectly on Insight and Python but when testing the .net environment, we received timeouts which (from looking at the sparse information in the Redis and Sentinel logs) appeared to be an SSL error.  My thought was that as we had not changed the .conf files that I needed to set on the following in the sentinel.conf:
sentinel resolve-hostnames yes
sentinel announce-hostnames yes
sentinel announce-ip "aaaaa-bbbbbb.cccc.dddddd.com"
sentinel announce-port 26379

Once again all still works well on the other two clients but now we get the following from the .net environment:

Unhandled exception. System.InvalidOperationException: The configuration is invalid. Creating the instance for type IConfirmableCommandHandler<UpdateDealFromMidasCommand, UpdateDealFromMidasConfirmation> failed. The registered delegate for type IAsyncCacheManager threw an exception. Timeout performing ROLE (5000ms), inst: 0, qu: 0, qs: 0, aw: False, bw: CheckingForTimeout, rs: NotStarted, ws: Initializing, in: 510, serverEndpoint: 22.242.107.85:6379, mc: 1/1/0, mgr: 10 of 10 available, clientName: markets-deals-service-command-host-5cf449b7f7-pdwn7(SE.Redis-v2.6.48.48654), IOCP: (Busy=0,Free=1000,Min=1,Max=1000), WORKER: (Busy=0,Free=32767,Min=1,Max=32767), POOL: (Threads=8,QueuedItems=0,CompletedItems=1488), v: 2.6.48.48654 (Please take a look at this article for some common client-side issues that can cause timeouts: https://stackexchange.github.io/StackExchange.Redis/Timeouts)
10/28/2022 9:02:09 AM  ---> SimpleInjector.ActivationException: The registered delegate for type IAsyncCacheManager threw an exception. Timeout performing ROLE (5000ms), inst: 0, qu: 0, qs: 0, aw: False, bw: CheckingForTimeout, rs: NotStarted, ws: Initializing, in: 510, serverEndpoint: 22.242.107.85:6379, mc: 1/1/0, mgr: 10 of 10 available, clientName: markets-deals-service-command-host-5cf449b7f7-pdwn7(SE.Redis-v2.6.48.48654), IOCP: (Busy=0,Free=1000,Min=1,Max=1000), WORKER: (Busy=0,Free=32767,Min=1,Max=32767), POOL: (Threads=8,QueuedItems=0,CompletedItems=1488), v: 2.6.48.48654 (Please take a look at this article for some common client-side issues that can cause timeouts: https://stackexchange.github.io/StackExchange.Redis/Timeouts)
10/28/2022 9:02:09 AM  ---> StackExchange.Redis.RedisTimeoutException: Timeout performing ROLE (5000ms), inst: 0, qu: 0, qs: 0, aw: False, bw: CheckingForTimeout, rs: NotStarted, ws: Initializing, in: 510, serverEndpoint: 22.242.107.85:6379, mc: 1/1/0, mgr: 10 of 10 available, clientName: markets-deals-service-command-host-5cf449b7f7-pdwn7(SE.Redis-v2.6.48.48654), IOCP: (Busy=0,Free=1000,Min=1,Max=1000), WORKER: (Busy=0,Free=32767,Min=1,Max=32767), POOL: (Threads=8,QueuedItems=0,CompletedItems=1488), v: 2.6.48.48654 (Please take a look at this article for some common client-side issues that can cause timeouts: https://stackexchange.github.io/StackExchange.Redis/Timeouts)
10/28/2022 9:02:09 AM    at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server, T defaultValue) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1873
10/28/2022 9:02:09 AM    at StackExchange.Redis.RedisServer.ExecuteSync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server, T defaultValue) in /_/src/StackExchange.Redis/RedisServer.cs:line 742
10/28/2022 9:02:09 AM    at StackExchange.Redis.ConnectionMultiplexer.GetSentinelMasterConnection(ConfigurationOptions config, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.Sentinel.cs:line 233
10/28/2022 9:02:09 AM    at StackExchange.Redis.ConnectionMultiplexer.Connect(ConfigurationOptions configuration, TextWriter log) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 677
10/28/2022 9:02:09 AM    at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
10/28/2022 9:02:09 AM    at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
10/28/2022 9:02:09 AM    at System.Lazy`1.CreateValue()
10/28/2022 9:02:09 AM    at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
10/28/2022 9:02:09 AM    at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
10/28/2022 9:02:09 AM    at System.Lazy`1.CreateValue()
10/28/2022 9:02:09 AM    at FX.Trading.Core.Cache.RedisStore.<>c.<.cctor>b__0_0()
10/28/2022 9:02:09 AM    at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
10/28/2022 9:02:09 AM    at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
10/28/2022 9:02:09 AM    at System.Lazy`1.CreateValue()
10/28/2022 9:02:09 AM    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
10/28/2022 9:02:09 AM    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
10/28/2022 9:02:09 AM    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
10/28/2022 9:02:09 AM    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
10/28/2022 9:02:09 AM    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
10/28/2022 9:02:09 AM    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
10/28/2022 9:02:09 AM    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
10/28/2022 9:02:09 AM    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
10/28/2022 9:02:09 AM    at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
10/28/2022 9:02:09 AM    at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
10/28/2022 9:02:09 AM    at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
10/28/2022 9:02:09 AM    at lambda_method(Closure )
10/28/2022 9:02:09 AM    at lambda_method(Closure )
10/28/2022 9:02:09 AM    at SimpleInjector.Lifestyles.SingletonLifestyle.SingletonLifestyleRegistration`1.CreateInstance(Func`1 instanceCreator)
10/28/2022 9:02:09 AM    at SimpleInjector.Lifestyles.SingletonLifestyle.SingletonLifestyleRegistration`1.CreateInstanceWithNullCheck()
10/28/2022 9:02:09 AM    at SimpleInjector.Lifestyles.SingletonLifestyle.SingletonLifestyleRegistration`1.GetInterceptedInstance()
10/28/2022 9:02:09 AM    at SimpleInjector.Lifestyles.SingletonLifestyle.SingletonLifestyleRegistration`1.BuildExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpressionInternal()
10/28/2022 9:02:09 AM    at SimpleInjector.Internals.LazyEx`1.get_Value()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpression()
10/28/2022 9:02:09 AM    --- End of inner exception stack trace ---
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildConstructorParameters(ConstructorInfo constructor)
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildNewExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildTransientExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpressionInternal()
10/28/2022 9:02:09 AM    at SimpleInjector.Internals.LazyEx`1.get_Value()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildConstructorParameters(ConstructorInfo constructor)
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildNewExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildTransientExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpressionInternal()
10/28/2022 9:02:09 AM    at SimpleInjector.Internals.LazyEx`1.get_Value()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildConstructorParameters(ConstructorInfo constructor)
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildNewExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildTransientExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpressionInternal()
10/28/2022 9:02:09 AM    at SimpleInjector.Internals.LazyEx`1.get_Value()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildConstructorParameters(ConstructorInfo constructor)
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildNewExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildTransientExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpressionInternal()
10/28/2022 9:02:09 AM    at SimpleInjector.Internals.LazyEx`1.get_Value()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildConstructorParameters(ConstructorInfo constructor)
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildNewExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.Registration.BuildTransientExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpressionInternal()
10/28/2022 9:02:09 AM    at SimpleInjector.Internals.LazyEx`1.get_Value()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.BuildExpression()
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.VerifyExpressionBuilding()
10/28/2022 9:02:09 AM    --- End of inner exception stack trace ---
10/28/2022 9:02:09 AM    at SimpleInjector.InstanceProducer.VerifyExpressionBuilding()
10/28/2022 9:02:09 AM    at SimpleInjector.Container.VerifyThatAllExpressionsCanBeBuilt(InstanceProducer[] producersToVerify)
10/28/2022 9:02:09 AM    at SimpleInjector.Container.VerifyThatAllExpressionsCanBeBuilt()
10/28/2022 9:02:09 AM    at SimpleInjector.Container.VerifyInternal(Boolean suppressLifestyleMismatchVerification)
10/28/2022 9:02:09 AM    at SimpleInjector.Container.Verify(VerificationOption option)
10/28/2022 9:02:09 AM    at Threshold.Cqrs.Configurator.CqrsHostScopes.CqrsHostScope`2..ctor(IConfigurationRoot configuration, Action`2 nonCqrsConfigurator, Action`1 contextValidationAction, Assembly[] cqrsAssemblies)
10/28/2022 9:02:09 AM    at Markets.Deals.CommandHost.Program.NewHost(IConfigurationRoot configuration) in /workspace/20/s/Solution/Markets.Deals.CommandHost/Program.cs:line 176
10/28/2022 9:02:09 AM    at Markets.Deals.CommandHost.Program.Main(String[] args) in /workspace/20/s/Solution/Markets.Deals.CommandHost/Program.cs:line 133
10/28/2022 9:02:09 AM 

The above from the client a docker image and at the Redis end we see: 

596300:M 28 Oct 2022 06:40:21.165 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:40:21.165 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:41:37.321 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:41:37.321 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:43:19.210 # Error accepting a client connection: Broken pipe
596300:M 28 Oct 2022 06:43:19.211 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:43:53.403 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.403 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.403 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.403 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.403 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.403 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.403 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.403 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:43:53.403 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.404 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.404 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.404 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.404 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.405 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.405 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.405 # Error accepting a client connection: (null) (conn: fd=7)
596300:M 28 Oct 2022 06:43:53.405 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:44:07.665 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:47:03.530 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:47:03.530 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:48:48.148 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:48:48.149 # Error accepting a client connection: Broken pipe
596300:M 28 Oct 2022 06:49:37.463 # Error accepting a client connection: (null)
596300:M 28 Oct 2022 06:49:37.464 # Error accepting a client connection: (null)

Sentinel shows:

597548:X 27 Oct 2022 13:35:10.914 # Error accepting a client connection: error:1408F10B:SSL routines:ssl3_get_record:wrong version number
604043:X 27 Oct 2022 15:37:54.323 # Error accepting a client connection: error:1408F10B:SSL routines:ssl3_get_record:wrong version number
604043:X 27 Oct 2022 15:37:55.136 # Error accepting a client connection: error:1408F10B:SSL routines:ssl3_get_record:wrong version number
604043:X 27 Oct 2022 15:38:14.027 # Error accepting a client connection: error:1408F10B:SSL routines:ssl3_get_record:wrong version number
604043:X 27 Oct 2022 15:38:14.028 # Error accepting a client connection: error:1408F10B:SSL routines:ssl3_get_record:wrong version number
604043:X 27 Oct 2022 15:38:14.089 # Error accepting a client connection: error:1408F10B:SSL routines:ssl3_get_record:wrong version number
604043:X 27 Oct 2022 15:38:14.091 # Error accepting a client connection: error:1408F10B:SSL routines:ssl3_get_record:wrong version number
604043:X 27 Oct 2022 15:38:14.118 # Error accepting a client connection: error:1408F10B:SSL routines:ssl3_get_record:wrong version number
604043:X 27 Oct 2022 15:38:14.118 # Error accepting a client connection: error:1408F10B:SSL routines:ssl3_get_record:wrong version number
604043:X 27 Oct 2022 15:38:14.163 # Error accepting a client connection: error:1408F10B:SSL routines:ssl3_get_record:wrong version number
604043:X 27 Oct 2022 15:38:14.165 # Error accepting a client connection: error:1408F10B:SSL routines:ssl3_get_record:wrong version number

Any help appreciated and I am just hoping that this is just something missing from my end.  As I mentioned though, the puzzling thing for me is that everything works find from the other two platforms.  Could it possibly be a "permissiveness" from Stack exchange parameter?

**Comment by NickCraver** [comment]
What does your connection config/string look like? The SSL3 error makes me think you may need to specify which protocols to use (e.g. TLS 1.2) and that it wasn't an issue before using ciphers incompatible with SSL3 for example..so it'll seem unrelated, but that's how it manifests.

**Comment by opencols** [comment]
Thank you Nick.  I should have included that originally.  Will wait for the apps guy on this.

Regards

John

**Comment by opencols** [comment]
Hi Nick,
Firstly, thanks again for coming back to us on this issue.  It has been a mission but after your reply we looked at the connection string (parameters) and while we are still putting the implementation through its paces, it appears that the sslHost parameter needed to be filled in with the one of the node’s urls.  In this case we made it of the current master.  Once we did that, the application worked.  Further, when doing a hard drop it continued to work, very much like Redis Insight (2) where it also continues to work after a drop although only one sentinel node has been given i.e., it picks another one automatically.

As an aside, it appears that pyRedis plugs the url in based upon the sentinel connection string.  That is a guess on my part.

I have left the “announce” and resolve parameters which at least seems to do no hard but as mentioned we are still “break” testing.  We restricted redis and Sentinel to TLSv1.2 and did the same in the Stack Exchange.  We will now gradually reinstate TLSv1.3.

Hopefully this is not an empty promise on my part, but I will try and share a write up on this experience and the Ansible script used to create the TLS based Redis instances…  Hopefully nobody will find my very basic ansible skills too funny 😊.

Once again thanks very Much

Regards

John
